src/libraries/Native/Unix/System.Globalization.Native/pal_timeZoneInfo.c

```diff
@@ -80,3 +58,283 @@ int32_t GlobalizationNative_IanaIdToWindowsId(const UChar* ianaId, UChar* window
+            return;
+        }
+
+        // Make a UChar[] version of the test time zone id for use in the API calls.
+        u_uastrcpy(testTimeZoneId, testId);
+
+        // Get the standard name from the test time zone.
+        GetTimeZoneDisplayName_FromCalendar(locale, testTimeZoneId, timestamp, UCAL_STANDARD, testDisplayName, DISPLAY_NAME_LENGTH, err);
+        if (U_FAILURE(*err))
+        {
+            // Failed, but keep trying through the rest of the loop in case the failure is specific to this test zone.
+            continue;
+        }
+
+        // See if the test time zone has a different standard name.
+        if (u_strcmp(testDisplayName, standardName) != 0)
+        {
+            // It has a different standard name. We can't use it.
+            continue;
+        }
+
+        // Get the generic name from the test time zone.
+        GetTimeZoneDisplayName_FromPattern(locale, testTimeZoneId, timestamp, GENERIC_PATTERN_UCHAR, testDisplayName, DISPLAY_NAME_LENGTH, err);
+        if (U_FAILURE(*err))
+        {
+            // Failed, but keep trying through the rest of the loop in case the failure is specific to this test zone.
+            continue;
+        }
+
+        // See if the test time zone has a longer (or same size) generic name.
+        if (u_strlen(testDisplayName) >= originalGenericNameActualLength)
+        {
+            // The test time zone's generic name isn't any shorter than the one we already have.
+            continue;
+        }
+
+        // We probably have found a better generic name.  But just to be safe, make sure the test zone isn't
+        // using a generic name that is specific to a particular location.  For example, "Antarctica/Troll"
+        // uses "Troll Time" as a generic name, but "Greenwich Mean Time" as a standard name.  We don't
+        // want other zones that use "Greenwich Mean Time" to be labeled as "Troll Time".
+
+        GetTimeZoneDisplayName_FromPattern(locale, testTimeZoneId, timestamp, GENERIC_LOCATION_PATTERN_UCHAR, testDisplayName2, DISPLAY_NAME_LENGTH, err);
+        if (U_FAILURE(*err))
+        {
+            // Failed, but keep trying through the rest of the loop in case the failure is specific to this test zone.
+            continue;
+        }
+
+        if (u_strcmp(testDisplayName, testDisplayName2) != 0)
+        {
+            // We have found a better generic name.  Use it.
+            u_strcpy(genericName, testDisplayName);
+            break;
+        }
+    }
+
+    uenum_close(pEnum);
+    ucal_close(calendar);
+}
+
+/*
+Gets the localized display name that is currently in effect for the specified time zone.
+*/
+ResultCode GlobalizationNative_GetTimeZoneDisplayName(const UChar* localeName, const UChar* timeZoneId, TimeZoneDisplayNameType type, UChar* result, int32_t resultLength)
+{
+    UErrorCode err = U_ZERO_ERROR;
+    char locale[ULOC_FULLNAME_CAPACITY];
+    GetLocale(localeName, locale, ULOC_FULLNAME_CAPACITY, false, &err);
```

Comment:
Should we check err after this call?

Reply:
We usually don't check for that in the other places. Here is some example https://github.com/dotnet/runtime/blob/dc5c9099b2c4ebcc74f8f40a5de5c4733ac79bfa/src/libraries/Native/Unix/System.Globalization.Native/pal_locale.c#L201

Reply:
I'll go ahead and add the check.  It's not expensive.